src/computability/turing_machine.lean

```diff
@@ -790,213 +1268,203 @@ def tr_normal : stmt₁ → stmt'
   stmt.branch (λ _ s, p a s) (tr_normal q₁) (tr_normal q₂)
 | (stmt.goto l)           := read $ λ a,
   stmt.goto (λ _ s, Λ'.normal (l a s))
-| stmt.halt               := move dir.right $ move dir.left $ stmt.halt
+| stmt.halt               := stmt.halt
 
-def tr_tape' (L R : list Γ) : tape bool :=
-tape.mk'
-  (L.bind (λ x, (enc x).to_list.reverse))
-  (R.bind (λ x, (enc x).to_list) ++ [default _])
+theorem step_aux_move (d q v T) :
+  step_aux (move d q) v T =
+  step_aux q v ((tape.move d)^[n] T) :=
+begin
+  suffices : ∀ i,
+    step_aux (stmt.move d^[i] q) v T =
+    step_aux q v (tape.move d^[i] T), from this n,
+  intro, induction i with i IH generalizing T, {refl},
+  rw [nat.iterate_succ', step_aux, IH, ← nat.iterate_succ]
+end
 
-def tr_tape : tape Γ → tape bool
-| (a, L, R) := tr_tape' L (a :: R)
+theorem supports_stmt_move {S d q} :
+  supports_stmt S (move d q) = supports_stmt S q :=
+suffices ∀ {i}, supports_stmt S (stmt.move d^[i] q) = _, from this,
+by intro; induction i generalizing q; simp only [*, nat.iterate]; refl
+
+theorem supports_stmt_write {S l q} :
+  supports_stmt S (write l q) = supports_stmt S q :=
+by induction l with a l IH; simp only [write, supports_stmt, *]
 
-theorem tr_tape_drop_right : ∀ R : list Γ,
-  list.drop n (R.bind (λ x, (enc x).to_list)) =
-  R.tail.bind (λ x, (enc x).to_list)
-| []     := list.drop_nil _
-| (a::R) := list.drop_left' (enc a).2
+theorem supports_stmt_read {S} : ∀ {f : Γ → stmt'},
+  (∀ a, supports_stmt S (f a)) → supports_stmt S (read f) :=
+suffices ∀ i (f : vector bool i → stmt'),
+  (∀ v, supports_stmt S (f v)) → supports_stmt S (read_aux i f),
+from λ f hf, this n _ (by intro; simp only [supports_stmt_move, hf]),
+λ i f hf, begin
+  induction i with i IH, {exact hf _},
+  split; apply IH; intro; apply hf,
+end
 
-parameters (enc0 : enc (default _) = vector.repeat ff n)
+parameter (enc0 : enc (default _) = vector.repeat ff n)
```

Comment:
We generally avoid `parameter` and use `variable` instead.  As far as I remember, parameters will also be gone in Lean 4.

Reply:
I tried, it's not better. I thought you could get close using e.g.
```lean
local notation `cfg` := cfg Γ Λ
```
but this has other side effects. For example, they apparently mess with re-binding declarations:
```

namespace ex1
section
parameter (A : Type)

def foo := A

parameter {A}

def bar (f : foo) : foo := f
end

#check @bar -- bar : Π {A : Type}, foo A → foo A
end ex1

namespace ex2
section
variable (A : Type)

def foo := A
local notation `foo` := foo A

variable {A}

def bar (f : foo) : foo := f
end

#check @bar -- bar : Π (A : Type), foo A → foo A
end ex2
```

Reply:
Ah, that's ugly but not too unexpected.  Then let's leave it like it is for now.